packages/prompts/src/index.ts

```diff
@@ -929,6 +929,85 @@ export const spinner = ({
 	};
 };
 
+export type ProgressOptions = SpinnerOptions & {
+	style?: 'light' | 'heavy' | 'block';
+	max?: number;
+	size?: number;
+};
+export interface ProgressResult {
+	start(msg?: string): void;
+	stop(msg?: string, code?: number): void;
+	message(msg?: string): void;
+	advance(step?: number, msg?: string): void;
+	readonly isCancelled: boolean;
+}
+
+export function progress({
+	style = 'heavy',
+	max = 100,
+	size = 40,
+	indicator = 'dots',
+	onCancel,
+	output = process.stdout,
+	cancelMessage,
+	errorMessage,
+	input,
+}: ProgressOptions = {}): ProgressResult {
+	const spin = spinner({
+		onCancel,
+		cancelMessage,
+		errorMessage,
+		indicator,
+		output,
+		input,
+	});
+	const progressChar: Record<NonNullable<ProgressOptions['style']>, string> = {
+		light: unicode ? '─' : '-',
+		heavy: unicode ? '━' : '=',
+		block: unicode ? '█' : '#',
+	};
+	let value = 0;
+	let previousMessage = '';
+
+	const _max = Math.max(1, max);
+	const _size = Math.max(1, size);
+
+	const activeStyle = (state: State) => {
+		switch (state) {
+			case 'initial':
+			case 'active':
+				return color.magenta;
+			case 'error':
+			case 'cancel':
+				return color.red;
+			case 'submit':
+				return color.green;
+		}
+		return color.magenta;
```

Comment:
typescript will detect this as unreachable since we have checked all possible `State` values

if you use `default` instead, typescript will be happy

```ts
default:
  return color.magenta;
```